File: weird_captcha_gym/shared_runtime/app/mechanics/horizon_relay.js
(() => {
  "use strict";
  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};

  const TAU = Math.PI * 2;
  const esc = (value) => String(value ?? "").replace(/[&<>\"']/g, (char) => ({"&":"&amp;","<":"&lt;",">":"&gt;",'"':"&quot;", "'":"&#39;"}[char]));
  const num = (value, fallback = 0) => Number.isFinite(Number(value)) ? Number(value) : fallback;
  const add = (a, b) => ({x: a.x + b.x, y: a.y + b.y, z: a.z + b.z});
  const sub = (a, b) => ({x: a.x - b.x, y: a.y - b.y, z: a.z - b.z});
  const dot = (a, b) => a.x * b.x + a.y * b.y + a.z * b.z;
  const length = (a) => Math.sqrt(dot(a, a));
  const unit = (a) => { const n = length(a) || 1; return {x: a.x / n, y: a.y / n, z: a.z / n}; };
  const deg = (radians) => radians * 180 / Math.PI;
  const rad = (degrees) => degrees * Math.PI / 180;

  function render(state, helpers) {
    if (window.horizonRelayModel?.timer) clearInterval(window.horizonRelayModel.timer);
    const interaction = String(state.control_condition?.interaction || state.interaction_mode || "full");
    const root = helpers.app;
    const planetRadius = num(state.planet?.radius, 150);
    const speed = num(state.rotation?.speed_deg_per_tick, 0.4);
    const tolerance = num(state.aim_tolerance_deg, 12);
    const stations = Array.isArray(state.stations) ? state.stations : [];
    const craft = Array.isArray(state.spacecraft) ? state.spacecraft : [];
    const model = {
      state, helpers, interaction, planetRadius, speed, tolerance, stations, craft,
      tick: 0, angle: 0, active: null, aim: null, selectedStation: null,
      progress: Object.fromEntries(craft.map((item) => [String(item.id), 0])),
      completed: new Set(), events: [], seq: 0, timer: null, dragging: null,
      submitted: false, failed: false,
    };

    document.body.dataset.mechanic = "horizon-relay";
    root.innerHTML = `
      <section class="hr-shell" data-interaction="${esc(interaction)}" data-challenge-id="${esc(state.challenge_id)}">
        <header class="hr-header">
          <div><p class="hr-kicker">DEEP SPACE NETWORK / HORIZON RELAY</p><h1>Keep the signal above the curve of Earth.</h1><p class="hr-prompt">${esc(state.prompt)}</p></div>
          <div class="hr-status"><span class="hr-live-dot"></span><b class="hr-status-text">TRACKING GLOBE</b><small>tick <b class="hr-tick">0</b></small></div>
        </header>
        <main class="hr-layout">
          <section class="hr-stage-card">
            <div class="hr-stage-top"><span>ORBITAL VIEW / OUTWARD IS UP</span><span class="hr-mode">${interaction === "full" ? "FULL DISH CONTROL" : "SIMPLIFIED RELAY DESK"}</span></div>
            <div class="hr-stage-wrap"><svg id="horizon-relay-stage" class="hr-stage" viewBox="0 0 800 560" role="img" aria-label="Three-dimensional rotating Earth with ground stations and spacecraft"></svg><div class="hr-stage-note">Front-facing markers are bright. Dim stations and spacecraft are on the far side; the pale horizon line is the local line-of-sight boundary.</div></div>
          </section>
          <aside class="hr-sidebar">
            <section class="hr-card hr-stations"><div class="hr-card-label">GROUND STATIONS <span class="hr-selected-station">NONE SELECTED</span></div><div class="hr-station-list"></div></section>
            <section class="hr-card hr-requests"><div class="hr-card-label">SPACECRAFT REQUESTS <span class="hr-delivery-count">0 / ${craft.length}</span></div><div class="hr-request-list"></div></section>
            <section class="hr-card hr-control-card"><div class="hr-card-label">${interaction === "full" ? "DIRECT DISH" : "PROXY RELAY DESK"}</div>${interaction === "full" ? `<p>Drag a bright station dish to a spacecraft marker. Drag again to refresh the pointing as the planet turns.</p>` : `<p>Select a station, choose a spacecraft above its horizon, then press HOLD LINK. The link stops when the local horizon or aim is lost.</p><div class="hr-proxy-row"><button id="hr-hold" type="button">HOLD LINK</button><button id="hr-stop" type="button">STOP</button></div>`}<div class="hr-link-readout">NO ACTIVE RELAY</div></section>
            <section class="hr-card hr-legend"><div class="hr-card-label">3D CHECK</div><p>Each station sits on the sphere's surface. A spacecraft is reachable only when its sightline leaves the station's outward hemisphere; a flat map would give the wrong handoff.</p><div class="hr-legend-diagram"><span class="hr-dot station-dot"></span><span>station</span><span class="hr-dot craft-dot"></span><span>spacecraft</span><span class="hr-dot horizon-dot"></span><span>above horizon</span></div></section>
          </aside>
        </main>
        <footer class="hr-footer"><div class="hr-readout" data-status="idle">SELECT A BRIGHT STATION AND READ ITS HORIZON.</div><button id="hr-certify" type="button">CERTIFY ALL RELAYS</button></footer>
      </section>`;

    const svg = root.querySelector("#horizon-relay-stage");
    const readout = root.querySelector(".hr-readout");
    const tickNode = root.querySelector(".hr-tick");
    const statusNode = root.querySelector(".hr-status-text");

    function stationAt(station, tick = model.tick) {
      const lat = rad(num(station.latitude));
      const lon = rad(num(station.longitude)) + rad(num(tick) * model.speed);
      return {x: planetRadius * Math.cos(lat) * Math.cos(lon), y: planetRadius * Math.sin(lat), z: planetRadius * Math.cos(lat) * Math.sin(lon)};
    }
    function craftAt(item) {
      const p = item.position || [0, 0, 0];
      return {x: num(p[0]), y: num(p[1]), z: num(p[2])};
    }
    function visible(station, item, tick = model.tick) {
      const s = stationAt(station, tick); const c = craftAt(item);
      return dot(s, sub(c, s)) > 0.0001;
    }
    function aimError(station, item, aim, tick = model.tick) {
      if (!aim) return Infinity;
      const target = sub(craftAt(item), stationAt(station, tick));
      const vector = {x: num(aim[0]), y: num(aim[1]), z: num(aim[2])};
      const denominator = length(target) * length(vector);
      if (!denominator) return Infinity;
      return deg(Math.acos(Math.max(-1, Math.min(1, dot(target, vector) / denominator))));
    }
    function project(point) {
      return {x: 400 + point.x * 0.90, y: 280 - point.y * 0.90, depth: point.z};
    }
    function craftById(id) { return craft.find((item) => String(item.id) === String(id)) || null; }
    function stationById(id) { return stations.find((item) => String(item.id) === String(id)) || null; }
    function say(message, status = "idle") { readout.dataset.status = status; readout.textContent = message; }
    function event(type, fields = {}) { model.seq += 1; model.events.push({seq: model.seq, type, tick: model.tick, ...fields}); }
    function visiblePairs() {
      return stations.flatMap((station) => craft.filter((item) => visible(station, item)).map((item) => [String(station.id), String(item.id)]));
    }
    function snapshot() {
      const stationPoints = Object.fromEntries(stations.map((station) => { const point = project(stationAt(station)); return [String(station.id), point]; }));
      const craftPoints = Object.fromEntries(craft.map((item) => { const point = project(craftAt(item)); return [String(item.id), point]; }));
      return {tick: model.tick, active: model.active ? {...model.active} : null, progress: {...model.progress}, completed: Array.from(model.completed), pairs: visiblePairs(), stationPoints, craftPoints};
    }
    model.snapshot = snapshot;
    model.visible = visible;
    model.project = project;

    function sceneMarkup() {
      const planet = `<circle cx="400" cy="280" r="${planetRadius * .9}" class="hr-planet"/><circle cx="400" cy="280" r="${planetRadius * .9 + 22}" class="hr-atmosphere"/><ellipse cx="400" cy="280" rx="${planetRadius * .9}" ry="${planetRadius * .9 * .31}" class="hr-latitude"/><ellipse cx="400" cy="280" rx="${planetRadius * .9 * .31}" ry="${planetRadius * .9}" class="hr-longitude"/><path d="M 266 280 Q 400 212 534 280 Q 400 348 266 280" class="hr-horizon-line"/>`;
      const craftMarkup = craft.map((item) => {
        const p = project(craftAt(item)); const stationVisible = stations.some((station) => visible(station, item)); const done = model.completed.has(String(item.id));
        const progress = model.progress[item.id] || 0; const required = num(item.required_ticks, 1);
        return `<g class="hr-craft ${stationVisible ? "is-reachable" : "is-hidden"} ${done ? "is-done" : ""}" data-craft="${esc(item.id)}" transform="translate(${p.x.toFixed(2)} ${p.y.toFixed(2)})"><circle r="26" class="hr-craft-halo"/><circle r="13" class="hr-craft-body" fill="${esc(item.color)}"/><text y="4" text-anchor="middle" class="hr-craft-icon">${esc(item.icon)}</text><text y="-25" text-anchor="middle" class="hr-craft-label">${esc(item.label)}</text><rect x="-28" y="31" width="56" height="5" rx="2" class="hr-progress-bg"/><rect x="-28" y="31" width="${(56 * Math.min(1, progress / required)).toFixed(2)}" height="5" rx="2" class="hr-progress"/></g>`;
      }).join("");
      const stationMarkup = stations.map((station) => {
        const p3 = stationAt(station); const p = project(p3); const front = p3.z >= -4; const selected = model.selectedStation === String(station.id); const active = model.active?.stationId === String(station.id);
        return `<g class="hr-station ${front ? "is-front" : "is-back"} ${selected ? "is-selected" : ""} ${active ? "is-active" : ""}" data-station="${esc(station.id)}" transform="translate(${p.x.toFixed(2)} ${p.y.toFixed(2)})"><circle r="${active ? 22 : 18}" class="hr-station-ring"/><path d="M -10 7 L 0 -13 L 10 7 Z" class="hr-dish" fill="${esc(station.color)}"/><circle r="5" class="hr-station-core" fill="${esc(station.color)}"/><text y="-26" text-anchor="middle" class="hr-station-label">${esc(station.label)}</text></g>`;
      }).join("");
      let beam = "";
      if (model.active) {
        const station = stationById(model.active.stationId); const target = craftById(model.active.craftId);
        if (station && target) { const a = project(stationAt(station)); const b = project(craftAt(target)); beam = `<path d="M ${a.x.toFixed(2)} ${a.y.toFixed(2)} L ${b.x.toFixed(2)} ${b.y.toFixed(2)}" class="hr-beam"/>`; }
      }
      const ghost = model.dragging ? `<path d="M ${model.dragging.start.x.toFixed(2)} ${model.dragging.start.y.toFixed(2)} L ${model.dragging.current.x.toFixed(2)} ${model.dragging.current.y.toFixed(2)}" class="hr-ghost-beam"/>` : "";
      return `<defs><radialGradient id="hr-earth"><stop offset="0" stop-color="#2d7184"/><stop offset=".75" stop-color="#184a68"/><stop offset="1" stop-color="#102d4d"/></radialGradient><filter id="hr-glow"><feGaussianBlur stdDeviation="4"/></filter></defs><rect width="800" height="560" class="hr-space"/>${planet}<circle cx="400" cy="280" r="${planetRadius * .9 - 3}" fill="url(#hr-earth)" class="hr-earth-fill"/>${beam}${craftMarkup}${stationMarkup}${ghost}<text x="400" y="514" text-anchor="middle" class="hr-axis-caption">ROTATING Y AXIS · STATION HORIZON IS GEOMETRIC, NOT A MAP EDGE</text>`;
    }

    function renderScene() {
      svg.innerHTML = sceneMarkup();
      tickNode.textContent = String(model.tick);
      statusNode.textContent = model.active ? "RELAY ACTIVE" : "TRACKING GLOBE";
      statusNode.parentElement.classList.toggle("is-relaying", Boolean(model.active));
      root.querySelector(".hr-selected-station").textContent = model.selectedStation ? String(stationById(model.selectedStation)?.label || model.selectedStation) : "NONE SELECTED";
      root.querySelector(".hr-delivery-count").textContent = `${model.completed.size} / ${craft.length}`;
      root.querySelector(".hr-link-readout").textContent = model.active ? `LINK · ${stationById(model.active.stationId)?.label || "?"} → ${craftById(model.active.craftId)?.label || "?"}` : "NO ACTIVE RELAY";
      root.querySelector(".hr-station-list").innerHTML = stations.map((station) => {
        const pairs = craft.filter((item) => visible(station, item)).map((item) => item.label).join(" · ") || "none above horizon";
        const selected = model.selectedStation === String(station.id) ? "is-selected" : "";
        if (interaction === "full") return `<div class="hr-station-info ${selected}" data-station-info="${esc(station.id)}"><span class="hr-station-swatch" style="--swatch:${esc(station.color)}"></span><b>${esc(station.label)}</b><small>${esc(pairs)} · drag the globe dish</small></div>`;
        return `<button type="button" class="hr-station-button ${selected}" data-station-button="${esc(station.id)}"><span class="hr-station-swatch" style="--swatch:${esc(station.color)}"></span><b>${esc(station.label)}</b><small>${esc(pairs)}</small></button>`;
      }).join("");
      root.querySelector(".hr-request-list").innerHTML = craft.map((item) => {
        const done = model.completed.has(String(item.id)); const active = model.active?.craftId === String(item.id); const progress = model.progress[item.id] || 0;
        return `<div class="hr-request ${done ? "is-done" : ""} ${active ? "is-active" : ""}" data-craft-row="${esc(item.id)}"><div class="hr-request-top"><span class="hr-request-icon" style="--craft:${esc(item.color)}">${esc(item.icon)}</span><b>${esc(item.label)}</b><span>${progress}/${num(item.required_ticks)}</span></div><div class="hr-request-bar"><i style="width:${Math.min(100, 100 * progress / Math.max(1, num(item.required_ticks)))}%"></i></div>${interaction === "simplified" ? `<button type="button" class="hr-target-button" data-target-button="${esc(item.id)}" ${done ? "disabled" : ""}>AIM THIS CRAFT</button>` : `<small class="hr-craft-hint">${done ? "DELIVERED" : "drag a bright dish here"}</small>`}</div>`;
      }).join("");
      bindControls();
    }

    function bindControls() {
      root.querySelectorAll("[data-station-button]").forEach((button) => button.addEventListener("click", () => {
        const stationId = String(button.dataset.stationButton); if (model.active) event("link_stop", {input_source: "station_button"}); model.active = null; model.selectedStation = stationId; event("select_station", {station_id: stationId, input_source: "station_button"}); say(`${stationById(stationId).label} SELECTED · READ THE BRIGHT HORIZON TARGETS.`); renderScene();
      }));
      root.querySelectorAll("[data-target-button]").forEach((button) => button.addEventListener("click", () => {
        if (model.submitted) return;
        const station = stationById(model.selectedStation); const target = craftById(button.dataset.targetButton);
        if (!station || !target) { say("SELECT A GROUND STATION FIRST.", "error"); return; }
        if (!visible(station, target)) { say("THAT SPACECRAFT IS BELOW THIS STATION'S HORIZON.", "error"); return; }
        const aim = unit(sub(craftAt(target), stationAt(station)));
        model.aim = {stationId: String(station.id), craftId: String(target.id), vector: [aim.x, aim.y, aim.z]};
        event("aim", {station_id: String(station.id), craft_id: String(target.id), input_source: "target_button", aim_vector: model.aim.vector});
        say(`${station.label} DISH AIMED AT ${target.label} · PRESS HOLD LINK.`); renderScene();
      }));
      root.querySelector("#hr-hold")?.addEventListener("click", () => {
        if (model.submitted) return;
        if (!model.aim || !visible(stationById(model.aim.stationId), craftById(model.aim.craftId))) { say("AIM A VISIBLE SPACECRAFT BEFORE HOLDING.", "error"); return; }
        if (!model.active) { model.active = {stationId: model.aim.stationId, craftId: model.aim.craftId}; event("hold_start", {input_source: "hold_button", station_id: model.aim.stationId, craft_id: model.aim.craftId}); say("LINK HELD · WATCH THE HORIZON.", "active"); renderScene(); }
      });
      root.querySelector("#hr-stop")?.addEventListener("click", () => { if (model.active) { event("link_stop", {input_source: "stop_button"}); model.active = null; say("LINK STOPPED · CHOOSE A NEW AIM."); renderScene(); } });
      root.querySelector("#hr-certify")?.addEventListener("click", submit);
      svg.querySelectorAll(".hr-station").forEach((node) => {
        node.addEventListener("pointerdown", (eventObject) => {
          if (interaction !== "full" || model.submitted) return;
          const point = pointerPoint(eventObject); model.dragging = {stationId: String(node.dataset.station), start: point, current: point, moved: false}; svg.setPointerCapture?.(eventObject.pointerId); eventObject.preventDefault();
        });
      });
      if (!svg.dataset.eventsBound) {
        svg.dataset.eventsBound = "true";
        svg.addEventListener("pointermove", (eventObject) => { if (!model.dragging) return; const point = pointerPoint(eventObject); model.dragging.current = point; if (Math.hypot(point.x - model.dragging.start.x, point.y - model.dragging.start.y) > 8) model.dragging.moved = true; renderScene(); });
        svg.addEventListener("pointerup", (eventObject) => { if (!model.dragging) return; const drag = model.dragging; const point = pointerPoint(eventObject); model.dragging = null; const candidate = craft.find((item) => { const p = project(craftAt(item)); return Math.hypot(p.x - point.x, p.y - point.y) < 48; }); const target = candidate && !model.completed.has(String(candidate.id)) ? candidate : null; if (!drag.moved) { model.selectedStation = drag.stationId; event("select_station", {station_id: drag.stationId, input_source: "station_click"}); say(`${stationById(drag.stationId).label} SELECTED · DRAG ITS DISH TO A BRIGHT SPACECRAFT.`); renderScene(); return; } if (!target) { say(candidate ? "THAT SPACECRAFT IS ALREADY DELIVERED." : "DROP THE DISH ON A SPACECRAFT MARKER.", "error"); renderScene(); return; } const station = stationById(drag.stationId); if (!visible(station, target)) { say("THAT SPACECRAFT IS BEHIND THE PLANET FROM THIS STATION.", "error"); renderScene(); return; } const aim = unit(sub(craftAt(target), stationAt(station))); model.selectedStation = drag.stationId; model.aim = {stationId: drag.stationId, craftId: String(target.id), vector: [aim.x, aim.y, aim.z]}; model.active = {stationId: drag.stationId, craftId: String(target.id)}; event("aim", {station_id: drag.stationId, craft_id: String(target.id), input_source: "dish_drag", aim_vector: model.aim.vector, started: true}); say(`${station.label} → ${target.label} · RELAY ACTIVE · REFRESH THE AIM AS EARTH TURNS.`, "active"); renderScene(); });
        svg.addEventListener("pointercancel", () => { if (model.dragging) { model.dragging = null; renderScene(); } });
      }
    }

    function pointerPoint(eventObject) { const rect = svg.getBoundingClientRect(); return {x: (eventObject.clientX - rect.left) * 800 / rect.width, y: (eventObject.clientY - rect.top) * 560 / rect.height}; }

    function step() {
      if (model.submitted) return;
      model.tick += 1; model.angle = model.tick * model.speed;
      if (model.active) {
        const station = stationById(model.active.stationId); const target = craftById(model.active.craftId); const good = station && target && visible(station, target) && aimError(station, target, model.aim?.vector, model.tick) <= model.tolerance;
        if (good) {
          model.progress[target.id] += 1; event("sample", {station_id: model.active.stationId, craft_id: model.active.craftId, delivered_after: model.progress[target.id]});
          if (model.progress[target.id] >= num(target.required_ticks, 1)) { event("transfer_complete", {craft_id: model.active.craftId, delivered: model.progress[target.id]}); model.completed.add(String(target.id)); model.active = null; model.aim = null; say(`${target.label} DELIVERY COMPLETE · CHOOSE THE NEXT HORIZON.`, "passed"); }
        } else { event("link_break", {station_id: model.active.stationId, craft_id: model.active.craftId, reason: station && target && !visible(station, target) ? "horizon" : "aim"}); model.active = null; say("LINK LOST · THE GLOBE MOVED THE DISH OR SPACECRAFT BELOW THE HORIZON.", "error"); }
      }
      renderScene();
    }

    async function submit() {
      if (model.submitted) return;
      model.submitted = true; event("submit", {input_source: "certify_button"});
      const payload = {mechanic_id: state.mechanic_id, task_id: state.task_id, challenge_id: state.challenge_id, interaction_mode: interaction, events: model.events, completed: model.completed.size === craft.length};
      try {
        const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
        const outcome = await response.json();
        if (outcome.passed === true) { helpers.setReadout("PASS", "passed"); say("PASS · ALL SPACECRAFT RELAYS DELIVERED.", "passed"); root.querySelector(".hr-shell").classList.add("is-pass"); return; }
        if (outcome.state) { model.failed = true; render(outcome.state, helpers); helpers.setReadout("FAIL · NEW RELAY", "error"); return; }
        model.submitted = false; say("FAIL · RELAY LEDGER REJECTED.", "error");
      } catch (_error) { model.submitted = false; say("SUBMISSION UNAVAILABLE · RETRY.", "error"); }
    }

    renderScene();
    model.timer = setInterval(step, num(state.rotation?.tick_ms, 100));
    window.horizonRelayModel = model;
    say(interaction === "full" ? "DRAG A BRIGHT STATION DISH TO A SPACECRAFT." : "SELECT A BRIGHT STATION AND READ ITS HORIZON.");
  }

  window.WeirdCaptchaMechanics.horizon_relay = {rootSelector: ".hr-shell", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/horizon_relay.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "horizon_relay"


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": message}


def _rotate_y(point: list[float], angle_deg: float) -> tuple[float, float, float]:
    angle = math.radians(float(angle_deg))
    x, y, z = (float(value) for value in point)
    return (x * math.cos(angle) - z * math.sin(angle), y, x * math.sin(angle) + z * math.cos(angle))


def _station_at(station: dict[str, Any], tick: int, speed: float, radius: float) -> tuple[float, float, float]:
    lat = math.radians(float(station["latitude"]))
    lon = math.radians(float(station["longitude"])) + math.radians(float(tick) * speed)
    return (radius * math.cos(lat) * math.cos(lon), radius * math.sin(lat), radius * math.cos(lat) * math.sin(lon))


def _dot(first: tuple[float, float, float], second: tuple[float, float, float]) -> float:
    return sum(a * b for a, b in zip(first, second))


def _sub(first: tuple[float, float, float], second: tuple[float, float, float]) -> tuple[float, float, float]:
    return tuple(a - b for a, b in zip(first, second))


def _norm(value: tuple[float, float, float]) -> float:
    return math.sqrt(_dot(value, value))


def _visible(station: dict[str, Any], craft: dict[str, Any], tick: int, speed: float, radius: float) -> bool:
    station_point = _station_at(station, tick, speed, radius)
    craft_point = tuple(float(value) for value in craft["position"])
    line = _sub(craft_point, station_point)
    return _dot(station_point, line) > 1e-6


def _aim_error(station: dict[str, Any], craft: dict[str, Any], tick: int, speed: float, radius: float, aim: Any) -> float:
    try:
        vector = tuple(float(value) for value in aim)
    except (TypeError, ValueError):
        return math.inf
    station_point = _station_at(station, tick, speed, radius)
    craft_point = tuple(float(value) for value in craft["position"])
    target = _sub(craft_point, station_point)
    if _norm(vector) <= 1e-8 or _norm(target) <= 1e-8:
        return math.inf
    cosine = max(-1.0, min(1.0, _dot(vector, target) / (_norm(vector) * _norm(target))))
    return math.degrees(math.acos(cosine))


def _contract_matches(truth: dict[str, Any], public: dict[str, Any]) -> bool:
    fields = (
        "task_id", "challenge_id", "planet_radius", "tick_ms",
        "rotation_speed_deg_per_tick", "aim_tolerance_deg", "stations", "spacecraft",
        "control_condition", "interaction_mode",
    )
    return all(truth.get(field) == public.get(field) for field in fields)


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    if payload.get("mechanic_id") != MECHANIC_ID or truth.get("mechanic_id") != MECHANIC_ID or public.get("mechanic_id") != MECHANIC_ID:
        return _fail("mechanic mismatch")
    if payload.get("task_id") != truth.get("task_id") or payload.get("challenge_id") != truth.get("challenge_id"):
        return _fail("stale task or challenge")
    if not _contract_matches(truth, public):
        return _fail("public/private horizon relay geometry differs")
    condition = truth.get("control_condition")
    interaction = str((condition or {}).get("interaction") or truth.get("interaction_mode") or "")
    if interaction not in {"simplified", "full"} or payload.get("interaction_mode") != interaction:
        return _fail("submitted interaction surface does not match the selected relay contract")
    expected_sources = {
        "simplified": {"station": "station_button", "aim": "target_button", "hold": "hold_button"},
        "full": {"station": "station_click", "aim": "dish_drag"},
    }[interaction]
    events = payload.get("events")
    if not isinstance(events, list) or not 1 <= len(events) <= 60000:
        return _fail("relay transcript missing or outside limits")
    stations = {str(item["id"]): item for item in truth.get("stations") or [] if isinstance(item, dict)}
    spacecraft = {str(item["id"]): item for item in truth.get("spacecraft") or [] if isinstance(item, dict)}
    if not stations or not spacecraft:
        return _fail("relay world is empty")
    radius = float(truth.get("planet_radius") or 0)
    speed = float(truth.get("rotation_speed_deg_per_tick") or 0)
    tolerance = float(truth.get("aim_tolerance_deg") or 0)
    current_station: str | None = None
    current_craft: str | None = None
    aim: Any = None
    active = False
    last_sample_tick: int | None = None
    progress = {craft_id: 0 for craft_id in spacecraft}
    completed: list[str] = []
    submitted = False
    previous_seq = 0
    for event in events:
        if not isinstance(event, dict) or event.get("seq") != previous_seq + 1:
            return _fail("event sequence is not contiguous")
        previous_seq += 1
        kind = str(event.get("type") or "")
        try:
            tick = int(event.get("tick"))
        except (TypeError, ValueError):
            return _fail(f"event {previous_seq} has no integer tick")
        if tick < 0 or tick > 20000:
            return _fail("event tick is outside the finite relay clock")
        if kind == "select_station":
            if event.get("input_source") != expected_sources["station"] or str(event.get("station_id")) not in stations:
                return _fail("station selection used the wrong input surface")
            current_station = str(event["station_id"])
            if active:
                active = False
                last_sample_tick = None
        elif kind == "aim":
            station_id, craft_id = str(event.get("station_id")), str(event.get("craft_id"))
            if event.get("input_source") != expected_sources["aim"] or station_id not in stations or craft_id not in spacecraft:
                return _fail("dish aim used the wrong interaction input")
            if craft_id in completed:
                return _fail("dish aim reopened an already delivered spacecraft")
            if interaction == "simplified" and current_station != station_id:
                return _fail("target button aimed a station that was not selected")
            if interaction == "full" and event.get("input_source") != "dish_drag":
                return _fail("full relay aim was not a dish drag")
            if not _visible(stations[station_id], spacecraft[craft_id], tick, speed, radius):
                return _fail("dish was aimed at a spacecraft below the selected station horizon")
            error = _aim_error(stations[station_id], spacecraft[craft_id], tick, speed, radius, event.get("aim_vector"))
            if error > 2.0:
                return _fail(f"dish aim vector disagrees with the visible 3D target ({error:.2f} degrees)")
            current_station, current_craft, aim = station_id, craft_id, event.get("aim_vector")
            if interaction == "full":
                active = True
                last_sample_tick = tick
        elif kind == "hold_start":
            if interaction != "simplified" or event.get("input_source") != expected_sources["hold"]:
                return _fail("hold control does not match simplified relay input")
            if current_station not in stations or current_craft not in spacecraft or aim is None:
                return _fail("hold started without an aimed station and spacecraft")
            if not _visible(stations[current_station], spacecraft[current_craft], tick, speed, radius):
                return _fail("hold started below the station horizon")
            active = True
            last_sample_tick = tick
        elif kind == "sample":
            if not active or current_station not in stations or current_craft not in spacecraft or aim is None:
                return _fail("transfer sample occurred without an active link")
            if tick != int(last_sample_tick) + 1:
                return _fail("transfer samples skip or repeat task ticks")
            visible = _visible(stations[current_station], spacecraft[current_craft], tick, speed, radius)
            error = _aim_error(stations[current_station], spacecraft[current_craft], tick, speed, radius, aim)
            if not visible or error > tolerance:
                return _fail("browser reported transfer while the replayed line of sight or aim was invalid")
            expected_after = progress[current_craft] + 1
            if progress[current_craft] >= int(spacecraft[current_craft]["required_ticks"]):
                return _fail("transfer sample exceeds the requested amount")
            if event.get("station_id") != current_station or event.get("craft_id") != current_craft or event.get("delivered_after") != expected_after:
                return _fail("transfer progress does not match the replayed link")
            progress[current_craft] = expected_after
            last_sample_tick = tick
        elif kind == "transfer_complete":
            if not active or current_craft not in spacecraft or current_craft in completed or progress[current_craft] != int(spacecraft[current_craft]["required_ticks"]):
                return _fail("transfer was marked complete before its samples")
            if event.get("craft_id") != current_craft or event.get("delivered") != int(spacecraft[current_craft]["required_ticks"]):
                return _fail("completion names the wrong spacecraft")
            completed.append(current_craft)
            active = False
            last_sample_tick = None
            current_craft = None
        elif kind == "link_break":
            if not active or current_station not in stations or current_craft not in spacecraft:
                return _fail("link break occurred without an active link")
            visible = _visible(stations[current_station], spacecraft[current_craft], tick, speed, radius)
            error = _aim_error(stations[current_station], spacecraft[current_craft], tick, speed, radius, aim)
            if visible and error <= tolerance:
                return _fail("link break was reported while the link was still valid")
            active = False
            last_sample_tick = None
        elif kind == "link_stop":
            if not active:
                return _fail("duplicate link stop")
            active = False
            last_sample_tick = None
        elif kind == "submit":
            submitted = True
        else:
            return _fail(f"unknown relay event {kind!r}")
    expected_completed = set(spacecraft)
    passed = submitted and payload.get("completed") is True and set(completed) == expected_completed and len(completed) == len(expected_completed) and all(progress[key] == int(value["required_ticks"]) for key, value in spacecraft.items())
    return {
        "graded": True,
        "passed": passed,
        "feedback": f"delivered {len(completed)}/{len(spacecraft)} spacecraft requests; replayed 3D horizon visibility and dish aiming",
    }
